feat(NODE-6993): Add support for FreeBSD

tagattie opened the pull request:
### Description

This PR contains a set of changes necessary for building this node module on FreeBSD.

#### What is changing?

Add conditions for enabling build on FreeBSD, namely:
- change `OS=="linux"` to `OS=="linux" or OS=="freebsd"` in `binding.gyp`,
- add `defined(__FreeBSD__)` in `src/kerberos_common.h` so that the UNIX-related sources are used.

##### Is there new documentation needed for these changes?

No.

#### What is the motivation for this change?

With the proposed change, FreeBSD users can use this node module on their systems.

<!-- If this is a bug, it helps to describe the current behavior and a clear outline of the expected behavior -->
<!-- If this is a feature, it helps to describe the new use case enabled by this change -->

<!--
Contributors!
First of all, thank you so much!!
If you haven't already, it would greatly help the team review this work in a timely manner if you create a JIRA ticket to track this PR.
You can do that here: https://jira.mongodb.org/projects/NODE
-->

### Release Highlight

<!-- RELEASE_HIGHLIGHT_START -->

### Add support for FreeBSD

FreeBSD is now an unofficially supported platform. Thank you to [tagattie](https://github.com/tagattie) for the contribution.

<!-- RELEASE_HIGHLIGHT_END -->

### Double check the following

- [x] Ran `npm run check:lint` script
- [x] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [x] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [x] Changes are covered by tests
- [x] New TODOs have a related JIRA ticket

tagattie committed:
Add support for FreeBSD

addaleax commented:
> feat(NODE-6993)!:

Is this breaking users? I can't see anything in the PR itself that would do that

tagattie commented:
> Is this breaking users?

No, I don't think so. This PR is just for adding support building on FreeBSD. No existing users should be affected.

addaleax commented:
Okay, then I'll remove the `!` from the PR title 🙂

durran commented:
Hi @tagattie, thank you very much for the contribution. We generally would not want to release this without sanity checks in our CI that this is building properly. Would it be possible to mirror our own build actions on a FreeBSD VM? I see there is this project that could be of help: https://github.com/vmactions/freebsd-vm

You can also reference our existing actions here: https://github.com/mongodb-js/kerberos/blob/main/.github/workflows/build.yml

tagattie commented:
Hi @durran, thank you for considering the proposed changes for merging.

> Would it be possible to mirror our own build actions on a FreeBSD VM?

Of cource, I will try to add an action for FreeBSD. I'm on it.

tagattie committed:
Add GitHub Actions for FreeBSD build

tagattie committed:
Install node package before run

tagattie committed:
Install npm before run as well

Node and npm are provided as separate packages in FreeBSD.

tagattie committed:
Of course we need to install krb5 package as well

tagattie committed:
Add FreeBSD-specific include and library dirs

FreeBSD packages are installed into /usr/local/... by default. So we need to add those directories to header and library search paths.

tagattie committed:
libcom_err.so on FreeBSD has different major version than Linux

This causes runtime error when the module is compiled with kerberos_use_rtld==true.

tagattie committed:
Align step name with host_builds step

tagattie commented:
I have added "freebsd_builds" action to build.yml. With some tweaks to `binding.gyp` and `src/unix/kerberos_gss.cc`, it successfully builds on FreeBSD CI environment.

I wait for your review. Please let me know if further changes is necessary.

addaleax commented on binding.gyp:
```diff
@@ -78,14 +78,24 @@
+              '/usr/local/lib',
+            ]
+          },
+        }],
```
I don't think it's a big deal to list these here, but shouldn't they be part of the default compiler include path to begin with?

tagattie committed:
Utilize pkg-config to get include/library paths for compiler

ELDiablO59152 commented:
Thanks a lot @tagattie !
